### PR TITLE
refactor(css): attribute selectors match web counterparts

### PR DESCRIPTION
**Pull request by bundyo**
## PR Checklist

- [x] The PR title follows our guidelines: https://github.com/NativeScript/NativeScript/blob/master/CONTRIBUTING.md#commit-messages.
- [x] There is an issue for the bug/feature this PR is for. To avoid wasting your time, it's best to open a suggestion issue first and wait for approval before working on it.
- [x] You have signed the [CLA](http://www.nativescript.org/cla).
- [x] All existing tests are passing: https://github.com/NativeScript/NativeScript/blob/master/DevelopmentWorkflow.md#running-unit-tests.
- [ ] Tests for the changes are included - https://github.com/NativeScript/NativeScript/blob/master/WritingUnitTests.md.

## What is the current behavior?
Since the new Theme 2.x produces bigger CSS and uses Element selector styling, the matched elements on the page are more and performance suffers. There are demo projects in the linked issue.

## What is the new behavior?
Exclude ProxyViewContainer from the CSS selector matching/init. Optimize some of the matchers and the selector query. Rework the attribute selectors to use exact matching, instead of regular expressions (as is done in browsers).

Fixes/Implements/Closes NativeScript/theme#179.

## BREAKING CHANGES:

~~Classes set on ProxyViewContainer will probably not work anymore. We might consider moving them to the wrapped element - either automatically in core modules or by the user?~~

### CODE REVIEW

File: tns-core-modules/ui/styling/css-selector/css-selector.ts [hunk 1]
```diff
@@ -412,9 +416,8 @@ export function fromAstNodes(astRules: cssParser.Node[]): RuleSet[] {
     return (<cssParser.Rule[]>astRules.filter(isRule)).map(rule => {
         let declarations = rule.declarations.filter(isDeclaration).map(createDeclaration);
         let selectors = rule.selectors.map(createSelector);
-        let ruleset = new RuleSet(selectors, declarations);
 
-        return ruleset;
+        return new RuleSet(selectors, declarations);
```
[the remark this answers is not in the document]
Comment (bundyo) [reply tns-core-modules/ui/styling/css-selector/css-selector.ts]: Not really. :) We can return it.

[the remark this answers is not in the document]
Comment (NathanaelA) [reply tns-core-modules/ui/styling/css-selector/css-selector.ts]: Actually technically their can be a BIG performance reason.  It is very hard to know for sure; and it really depends on number of times this function is called...

The first method creates a named variable inside the local context; then pushes it on the stack for returning, which adds _additional_ work to GC.    

The second version should create a variable directly on the stack for returning as the engine now knows this doesn't need to be tracked as part of this function closure.   Eliminating one or more different items from having to tracked, calculated and GC'd...

So basically in a nutshell this should eliminate some additional GC work.      If this function is called often enough; it could actually make a major difference...

[the remark this answers is not in the document]
Comment (bundyo) [reply tns-core-modules/ui/styling/css-selector/css-selector.ts]: Depending on the number of rules it might be called 100 or 1000 times. Probably the impact won't be that high for a normal project though.

[the remark this answers is not in the document]
Comment (bundyo) [reply tns-core-modules/ui/styling/css-selector/css-selector.ts]: Depending on the number of rules it might be called 100 or 1000 times. Probably the impact won't be that high for a normal project though.

**Comment by MartoYankov** [comment]
@manoldonev Can you take a look at this? I couldn't find a scenario where there are significant performance gains (the projects mentioned had a different bug), but I think it's good even if only as a refactoring.

**Comment by edusperoni** [comment]
I'm working in allowing some layout properties to cascade from ProxyViewContainer to the children, so this breaking change will potentially make it impossible to implement something like CSS Grid. Can this breaking change be clarified or maybe worked around?

**Comment by bundyo** [comment]
The main issue turned out to be elsewhere, so this PR can be reduced to only get the attribute selectors match their web counterparts.

**Comment by manoldonev** [comment]
@edusperoni Can you elaborate on your concern about the `ProxyViewContainer` change? Until now we have not actually considered setting css classes on `ProxyViewContainer` as a meaningful supported scenario. Are you actually trying to set such css classes or you are concerned that cascading will not work? Based on the current change I think cascading should continue to work?

**Comment by edusperoni** [comment]
@manoldonev Please see the discussion in https://github.com/NativeScript/nativescript-angular/issues/1422#issuecomment-546062221

My current plan is, once we settle on the implementation details of forwarding properties to children, to do:

```html
<GridLayout class="grid-container">
  <my-component class="grid-header"></my-component>
  <my-component class="grid-left"></my-component>
  <my-component class="grid-content"></my-component>
  <my-component class="grid-footer"></my-component>
</GridLayout>
```

```css
.grid-container {
   grid-template-areas: 'header header' 'left content' 'footer footer';
}

.grid-header {
   grid-area: header;
}
.grid-left {
   grid-area: left;
}
.grid-content {
   grid-area: content;
}
.grid-footer {
   grid-area: footer;
}
```

GridLayout would then take the children styles and apply row/col to them as required. If we remove the ability for ProxyViewContainer to be matched and properly hold styles, we'll no longer be able to implement grid or add properties that will be forwarded to children (like row/col) in CSS
